src/connector/src/source/mqtt/enumerator/mod.rs

```diff
@@ -117,10 +119,18 @@ impl SplitEnumerator for MqttSplitEnumerator {
 
     async fn list_splits(&mut self) -> ConnectorResult<Vec<MqttSplit>> {
         if !self.connected.load(std::sync::atomic::Ordering::Relaxed) {
-            return Err(ConnectorError::from(MqttError(format!(
-                "Failed to connect to MQTT broker for topic {}",
-                self.topic
-            ))));
+            let start = std::time::Instant::now();
+            loop {
+                if self.connected.load(std::sync::atomic::Ordering::Relaxed) {
+                    break;
+                }
+
+                if start.elapsed().as_secs() > 10 {
+                    bail!("Failed to connect to mqtt broker");
+                }
+
+                tokio::time::sleep(std::time::Duration::from_millis(500)).await;
+            }
```

Comment:
regarding the comments https://github.com/risingwavelabs/risingwave/pull/18961#issuecomment-2422455694, I still have some concern about the wait time. Would you be able to move the logic in the tokio thread into this func?

Reply:
I don't fully understand exactly what you're suggesting, but I'd say it's one of:
1. Move the logic inside the tokio thread to the `list_splits`, spawn the thread in `new`
2. During the call to `list_splits` spawn a thread and then stop that thread when the topics have come in
3. During the call to `list_splits` spawn a new thread to listen to the incoming mqtt topics
4. During the call to `list_splits` spawn a new thread if one isn't already running to listen to the incoming mqtt topics

Option 3 would spawn a new thread on each enumerator tick, which doesn't seem desirable to me
Option 4 is functionally identical to what we have just with added state to track the existence of the thread; I don't really see what would be gained
Option 2 would risk missing topics that come in after `list_splits` returns and come in before the next `tick`
Option 1 would require some inter thread communication in order to `poll` the mqtt eventloop inside the spawned thread whilst having the logic in the `list_splits`. That seems to make it more complicated than desired

Please indicate if I've misunderstood

Reply:
> Option 2 would risk missing topics that come in after list_splits returns and come in before the next tick

Oh, I got it. Mqtt does not have persisted metadata so RisingWave needs a thread running in the background. The fix makes sense.